### ISSUE TITLE
Add preliminary OSL support

### ISSUE THREAD

**Issue by doug-walker** [AcademySoftwareFoundation/OpenColorIO#1435]
OpenShadingLanguage (OSL), which is also an ASWF project, has many similarities to the shading languages that OCIO's GPU renderer can currently generate such as GLSL and HLSL.  Therefore, for most ops, it would not be too difficult to provide the option to emit OSL as one of the supported languages.  This would be useful in the context of adding OCIO support to MaterialX, for example (another ASWF project!).

The most difficult issues are how to handle LUTs (which are represented as textures by the GPU path) and dynamic properties (which are handled as uniforms on the GPU).  We have been discussing ideas for how to handle these, but the proposal is that an initial implementation would throw if the color space conversion involves LUTs and that dynamic properties would be set to their initial values (in other words, a new Processor would need to be created to modify them).  Given that the needs for color management within the context of, for example a MaterialX shading network, tend to be less than the needs of a general purpose application, our sense is that these limitations might be acceptable for an initial implementation, but we are interested in feedback on this question.

For reference, tagging issue #261 here, since there is some overlap in some of the discussion around that one.

**Comment by lgritz** [comment]
For LUTs, depending on the size, they can be just arrays in the language (in the process of optimization and JITing, an array in OSL that is only written to in the process of its declaration/initialization will be turned into a static constant, and will not be filled in anew every time the shader runs). Larger LUTs can, of course, just be textures.

I'm not sure what you mean by dynamic properties / uniforms. Can you ELI5? (I should know more internal details and nomenclature of OCIO, I know, but in practice I tend to only directly touch the API calls I need to get a color processor and use it to transform arrays of colors.)

**Comment by doug-walker** [comment]
@lgritz, here's an explanation of dynamic properties ... For most OCIO Transforms, for example if you have a MatrixTransform, and want to change the matrix coefficients, the application would have to change them on the Transform object and then build a new Processor.  However, there are some use-cases where users want to be able to modify values using a slider in a UI and get real-time feedback.  For example, the ExposureContrast Transform is used to implement the real-time exposure and gamma adjustment used in a DCC viewport.  Having to recreate a Processor, export a shader, and upload that to the GPU for each micro move of the slider might result in an annoying amount of lag for the user.  So a few Transform types support the dynamic property interface which means that it is possible to edit values in an existing Processor.  That is, an application can basically request a handle to change a numeric value inside an existing Processor.  On the GPU side, these are mapped to uniforms.  So rather than uploading a new shader just to change one number, the application is just updating the value of a uniform in an existing shader program.

**Comment by lgritz** [comment]
I don't know how/if that's relevant to MaterialX.

But as far as OSL is concerned, you just make the matrix (or whatever) parameters to the shader itself. Now, ordinarily, those parameter values would get all constant folded as part of the JIT, but there is a way to say that they are not locked (in the source code, the shader param is declared with metadata `[[ int lockgeom=0 ]]`, and there's also a way to specify it for particular parameters through the APIs when you tell the ShadingSystem about that shader network. This is all easy to set up, if it's even needed at all.

**Comment by hodoulp** [comment]
@lgritz thanks for the hint.

**Comment by hodoulp** [comment]
To track questions/discussions on the OSL support by OCIO, here is the [discussion thread](https://lists.aswf.io/g/osl-dev/topic/ociov2_with_osl_native/84008441?p=,,,20,0,0,0::recentpostdate%2Fsticky,,,20,2,0,84008441)
